From b7983f6f1cec6e9177c1175d42a7a576e3dd9215 Mon Sep 17 00:00:00 2001
From: Noam Camus <ncamus@habana.ai>
Date: Sat, 04 Aug 2018 21:17:16 +0300
Subject: [PATCH] Add dali packets support fixup

Update flash_burn script to new flash layout
No more golden.
We switch between FW0 and FW1
---

diff --git a/board/habanalabs/common/rootfs_overlay/usr/sbin/flash_burn b/board/habanalabs/common/rootfs_overlay/usr/sbin/flash_burn
index f58ec0f..139c354 100755
--- a/board/habanalabs/common/rootfs_overlay/usr/sbin/flash_burn
+++ b/board/habanalabs/common/rootfs_overlay/usr/sbin/flash_burn
@@ -1,20 +1,22 @@
-#!/bin/sh
+#!/bin/sh -x
 
 if [ $# -eq 0 ]; then
-	echo "Usage: ${0} [-s] [-c <cksum>] <image>"
+	echo "Usage: ${0} [-c <cksum>] <image>"
+	echo "Usage: ${0} -s <next_id>"
 	echo "       burn <image> into flash sectors"
-	echo "       -s, skip golden flash sectors"
 	echo "       -c <cksum>, crc32 for <image> validation"
+	echo "       -s <next_id>, set <next_id> as valid image for next boot"
 	exit 0
 else
 	while [ $# -gt 0 ]; do
 		case ${1} in
-			-s)
-				IS_SAFE_BURN="yes"
-				shift
-			;;
 			-c)
 				CKSUM=`printf "%d" ${2}`
+				shift
+				shift
+			;;
+			-s)
+				NEXT_IMAGE=`printf "%d" ${2}`
 				shift
 				shift
 			;;
@@ -26,6 +28,18 @@
 	done
 fi
 
+if [ -n "${NEXT_IMAGE}" ]; then
+	echo "Update next image to fw-$((${NEXT_IMAGE}-1))"
+	IMG=/tmp/next-image.txt
+	if [ ${NEXT_IMAGE} -eq 1 ]; then
+		echo "11110000" > ${IMG}
+	else
+		echo "22220000" > ${IMG}
+	fi
+
+	flashcp ${IMG} /dev/mtd2
+	exit 0
+fi
 
 if [ -z "${INFILE}" ]; then
 	echo "Missing image argument"
@@ -42,34 +56,48 @@
 	fi
 fi
 
-DEVNUM=`mtdinfo				| grep "Count of MTD devices"	| tr -s ' ' | cut -d: -f2	| cut -d' ' -f2`
-FULL_DEV_LIST="$(seq 0 $((${DEVNUM}-1)))"
+# MTD Partitions:
+# 0 -- All
+# 1 -- BTL Primary
+# 2 -- Next Image
+# 3 -- FW 0
+# 4 -- FW 1
+STABLE_IMAGE_CHAR=0x`xxd -p -l1 -c1 /dev/mtd2`
+# '1' / '2' equal (ascii) 0x31 / 0x32
+if ![ ${STABLE_IMAGE_CHAR} -eq 0x31 ] && ![ ${STABLE_IMAGE_CHAR} -eq 0x32 ]; then
+	STABLE_IMAGE_CHAR=0x31
+fi
+STABLE_IMAGE_NUM=$((${STABLE_IMAGE_CHAR} - 0x31))
+MODIFIABLE_IMAGE_ID=$((2 - ${STABLE_IMAGE_NUM}))
+MODIFIABLE_MTD_ID=$((3 + ${MODIFIABLE_IMAGE_ID}))
+BTL_IMAGE=${INFILE}.mtd1
+NEXT_IMAGE=${INFILE}.mtd2
+FW_IMAGE=${INFILE}.mtd${MODIFIABLE_MTD_ID}
+
+dd status=none conv=sync bs=64K if=${INFILE} of=${BTL_IMAGE} count=1
+if [ ${MODIFIABLE_IMAGE_ID} -eq 0 ]; then
+	echo "22221111" > next-image.txt
+else
+	echo "11112222" > next-image.txt
+fi
+dd status=none conv=sync bs=64K if=next-image.txt of=${NEXT_IMAGE}
+dd status=none conv=sync bs=64K skip=8 if=${INFILE} of=${FW_IMAGE}
 
 function burn_image {
 	DEV=/dev/mtd${1}
 	NAME=`mtdinfo ${DEV}		| grep "Name" 			| tr -s ' ' | cut -d: -f2	| cut -d' ' -f2`
-	DEV_IS_SAFE=`mtdinfo ${DEV}	| grep "Name"			| grep " \[safe]"`
-	if [ -n "${IS_SAFE_BURN}" ] && [ -z "${DEV_IS_SAFE}" ]; then
-		return
-	fi
 
-	BLK_SIZE=`mtdinfo ${DEV}	| grep "Eraseblock size"	| tr -s ' ' | cut -d: -f2	| cut -d' ' -f2`
-	BLK_CNT=`mtdinfo ${DEV}		| grep "Amount of eraseblocks"	| tr -s ' ' | cut -d: -f2	| cut -d' ' -f2`
 	DEVID=`mtdinfo ${DEV}		| grep "major/minor"		| tr -s ' ' | cut -d: -f2-3	| cut -d' ' -f2`
-	if [ ! -e /sys/dev/char/${DEVID}/offset ]; then
-		return
-	fi
 	OFFSET=`cat /sys/dev/char/${DEVID}/offset`
 	OFFSET_HEX=`printf "0x%x" ${OFFSET}`
-	SKIP_CNT=$((${OFFSET}/${BLK_SIZE}))
 	IMG=${INFILE}.mtd${1}
 
-	dd status=none conv=sync bs=${BLK_SIZE} skip=${SKIP_CNT} if=${INFILE} of=${IMG} count=${BLK_CNT}
 	echo -n "    ${NAME} @ ${OFFSET_HEX}... "
-	#flash_erase -q -u ${DEV} 0 ${BLK_CNT}
 	flashcp ${IMG} ${DEV}
 
-	dd status=none conv=sync bs=${BLK_SIZE} if=${DEV} of=${IMG}.verify count=${BLK_CNT}
+	IMG_SIZE=`ls -l ${IMG}						| tr -s ' ' 			| cut -d' ' -f5`
+	BLK_CNT=$(((${IMG_SIZE} + 0x8000) / 0x10000))
+	dd status=none conv=sync bs=64K if=${DEV} of=${IMG}.verify count=${BLK_CNT}
 	VERIFY=`diff -q ${IMG} ${IMG}.verify`
 	if [ -n "${VERIFY}" ]; then
 		echo "FAIL"
@@ -80,6 +108,6 @@
 }
 
 echo "Updating MTD Partitions:"
-for i in ${FULL_DEV_LIST}; do
-	burn_image ${i}
-done
+burn_image 1
+burn_image 2
+burn_image ${MODIFIABLE_MTD_ID}
